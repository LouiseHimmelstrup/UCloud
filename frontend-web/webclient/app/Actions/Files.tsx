import { Cloud } from "../../authentication/SDUCloudObject";
import {
    RECEIVE_FILES,
    UPDATE_FILES,
    SET_FILES_LOADING,
    UPDATE_PATH,
    SET_FILES_SORTING_COLUMN,
    FILE_SELECTOR_SHOWN,
    SET_FILE_SELECTOR_LOADING,
    RECEIVE_FILE_SELECTOR_FILES,
    SET_FILE_SELECTOR_CALLBACK,
    SET_DISALLOWED_PATHS,
    SET_CREATING_FOLDER,
    SET_EDITING_FILE,
    RESET_FOLDER_EDITING
} from "../Reducers/Files";
import { sortFilesByTypeAndName, failureNotification } from "../UtilityFunctions";
import { Page, emptyPage, File } from "../types/types";
import { SortOrder, SortBy } from "../SiteComponents/Files/Files";

/**
* Creates a promise to fetch files. Sorts the files based on sorting function passed,
* and implicitly sets {filesLoading} to false in the reducer when the files are fetched.
* @param {string} path is the path of the folder being queried
* @param {number} itemsPerPage number of items to be fetched
* @param {Page<File>} page number of the page to be fetched
*/
export const fetchFiles = (path: string, itemsPerPage: number, page: number, order: SortOrder, sortBy: SortBy) =>
    Cloud.get(`files?path=${path}&itemsPerPage=${itemsPerPage}&page=${page}&order=${order}&sortBy=${sortBy}`).then(({ response }) =>
        receiveFiles(response, path, order, sortBy)
    ).catch(() => {
        failureNotification("An error occurred fetching files for this folder.");
        return receiveFiles(emptyPage, path, order, sortBy);
    });

/**
* Updates the files stored. 
* Intended for use when sorting the files, checking or favoriting, for instance.
* @param {Page<File>} page contains the currently held page with modifications made to the files client-side
*/
export const updateFiles = (page: Page<File>) => ({
    type: UPDATE_FILES,
    page
});

/**
 * Sets whether or not the component is loading
 * @param {boolean} loading 
 */
export const setLoading = (loading: boolean) => ({
    type: SET_FILES_LOADING,
    loading
});

/**
 * Updates the path currently held intended for the files/fileinfo components.
 * @param {string} path - The path to store
 */
export const updatePath = (path: string) => ({
    type: UPDATE_PATH,
    path
});

/**
 * The function used for the actual receiving the files, rather than the promise
 * @param {Page<File>} page - Contains the page
 * @param {string} path - The path the files were retrieved from
 * @param {SortOrder} sortOrder - The order in which the files were sorted
 * @param {SortBy} sortBy - the value the sorting was based on
 */
const receiveFiles = (page: Page<File>, path: string, sortOrder: SortOrder, sortBy: SortBy) => {
    page.items.forEach((f) => f.isChecked = false);
    return {
        type: RECEIVE_FILES,
        page,
        path,
        sortOrder,
        sortBy
    }
};


export const setSortingColumn = (index, name) => ({
    type: SET_FILES_SORTING_COLUMN,
    index,
    name
});

export const fileSelectorShown = (state) => ({
    type: FILE_SELECTOR_SHOWN,
    state
});

export const receiveFileSelectorFiles = (files, path) => ({
    type: RECEIVE_FILE_SELECTOR_FILES,
    files,
    path
});

export const fetchPageFromPath = (path: string, itemsPerPage: number, order: SortOrder, sortBy: SortBy) =>
    Cloud.get(`files/lookup?path=${path}&itemsPerPage=${itemsPerPage}&order=${order}&sortBy=${sortBy}`)
        .then(({ response }) => receiveFiles(response, path, order, sortBy)).catch(() => {
            failureNotification(`An error occured fetching the page for file ${path}`);
            return { type: "ERROR" };
        }
    );

// FIXME add pagination to FileSelector and rewrite this:
export const fetchFileselectorFiles = (path: string, page: number, itemsPerPage: number) =>
    Cloud.get(`files?path=${path}&page=${page}&itemsPerPage=${itemsPerPage}`).then(({ response }) => {
        let files = response.items;
        files.forEach(file => file.isChecked = false);
        files = sortFilesByTypeAndName(files, true);
        return receiveFileSelectorFiles(files, path);
    }).catch(() => {
        failureNotification("An error occurred when fetching files for fileselection");
        return { type: "ERROR" }; // FIXME Will end up in default. Should have case for this
    });

export const setFileSelectorLoading = () => ({
    type: SET_FILE_SELECTOR_LOADING
})

export const setDisallowedPaths = (paths) => ({
    type: SET_DISALLOWED_PATHS,
    paths
});

export const setFileSelectorCallback = (callback) => ({
    type: SET_FILE_SELECTOR_CALLBACK,
    callback
});

export const setEditingFile = (editFileIndex: number) => ({
    type: SET_EDITING_FILE,
    editFileIndex
});

export const setCreatingFolder = (creatingFolder: boolean) => ({
    type: SET_CREATING_FOLDER,
    creatingFolder
});

export const resetFolderEditing = () => ({
    type: RESET_FOLDER_EDITING
})