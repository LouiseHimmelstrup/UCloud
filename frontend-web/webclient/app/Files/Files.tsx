import * as React from "react";
import { connect } from "react-redux";
import { Cloud } from "Authentication/SDUCloudObject";
import { Link } from "react-router-dom";
import { Dropdown, Button, Icon, Table, Header, Input, Grid, Responsive, Checkbox, Divider } from "semantic-ui-react";
import { setUploaderVisible, setUploaderCallback } from "Uploader/Redux/UploaderActions";
import { dateToString } from "Utilities/DateUtilities";
import * as Pagination from "Pagination";
import { BreadCrumbs } from "Breadcrumbs/Breadcrumbs";
import * as UF from "UtilityFunctions";
import { KeyCode, ReduxObject } from "DefaultObjects";
import * as Actions from "./Redux/FilesActions";
import { updatePageTitle } from "Navigation/Redux/StatusActions";
import { FileSelectorModal } from "./FileSelector";
import { FileIcon } from "UtilityComponents";
import { Page } from "Types";
import {
    FilesProps, SortBy, SortOrder, FilesStateProps, FilesOperations, File, FilesTableHeaderProps, FilenameAndIconsProps,
    FileOptionsProps, FilesTableProps, SortByDropdownProps, FileOperation, ContextButtonsProps, Operation, ContextBarProps,
    PredicatedOperation, ResponsiveTableColumnProps
} from ".";
import { setPrioritizedSearch } from "Navigation/Redux/HeaderActions";
import {
    startRenamingFiles, AllFileOperations, newMockFolder, isInvalidPathName, favoriteFileFromPage, getFilenameFromPath,
    isProject, toFileText, getParentPath, isDirectory, moveFile, createFolder, previewSupportedExtension
} from "Utilities/FileUtilities";
import { Dispatch } from "redux";

class Files extends React.Component<FilesProps> {

    componentDidMount() {
        const { page, fetchFiles, sortOrder, sortBy, history, setPageTitle, prioritizeFileSearch } = this.props;
        setPageTitle();
        prioritizeFileSearch();
        this.props.setUploaderCallback(
            (path: string) => fetchFiles(path, page.itemsPerPage, page.pageNumber, sortOrder, sortBy)
        );
        if (!this.urlPath) { history.push(`/files/${Cloud.homeFolder}/`); }
        else { fetchFiles(this.urlPath, page.itemsPerPage, page.pageNumber, sortOrder, sortBy); }
    }

    get urlPath(): string { return this.props.match.params[0]; }

    newFolder() {
        let { page, updateFiles } = this.props;
        page.items = [newMockFolder()].concat([...page.items.filter(it => !it.isMockFolder)]);
        updateFiles(page);
    }

    onRenameFile = (key: number, file: File, name: string) => {
        const { path, fetchPageFromPath, updateFiles, page } = this.props;
        if (key === KeyCode.ESC) {
            const item = page.items.find(f => f.path === file.path);
            if (item !== undefined) item.beingRenamed = false;
            page.items = page.items.filter(it => !it.isMockFolder);
            updateFiles(page);
        } else if (key === KeyCode.ENTER) {
            const fileNames = page.items.map((it) => getFilenameFromPath(it.path));
            if (isInvalidPathName(name, fileNames)) return;
            const fullPath = `${UF.addTrailingSlash(path)}${name}`;
            if (file.isMockFolder) {
                createFolder(fullPath, Cloud,
                    () => fetchPageFromPath(fullPath, page.itemsPerPage, this.props.sortOrder, this.props.sortBy));
            } else {
                moveFile(file.path, fullPath, Cloud,
                    () => fetchPageFromPath(fullPath, page.itemsPerPage, this.props.sortOrder, this.props.sortBy));
            }
        }
    }

    shouldComponentUpdate(nextProps, _nextState): boolean {
        const { fetchFiles, page, loading, sortOrder, sortBy } = this.props;
        if (nextProps.path !== nextProps.match.params[0] && !loading) {
            fetchFiles(nextProps.match.params[0], page.itemsPerPage, 0, sortOrder, sortBy);
        }
        return true;
    }

    render() {
        const { page, path, loading, history, fetchFiles, checkFile, updateFiles, sortBy, sortOrder, leftSortingColumn,
            rightSortingColumn, setDisallowedPaths, setFileSelectorCallback, showFileSelector, error } = this.props;
        const selectedFiles = page.items.filter(file => file.isChecked);
        const checkbox = (<Checkbox
            className="hidden-checkbox checkbox-margin"
            onClick={(_, d) => this.props.checkAllFiles(!!d.checked)}
            checked={page.items.length === selectedFiles.length && page.items.length > 0}
            indeterminate={selectedFiles.length < page.items.length && selectedFiles.length > 0}
            onChange={(e) => e.stopPropagation()}
        />);
        const refetch = () => fetchFiles(path, page.itemsPerPage, page.pageNumber, sortOrder, sortBy);
        const navigate = (path: string) => history.push(`/files/${path}`);
        const fetchPageFromPath = (path: string) => {
            this.props.fetchPageFromPath(path, page.itemsPerPage, sortOrder, sortBy);
            this.props.updatePath(getParentPath(path));
            navigate(getParentPath(path));
        };
        const fileSelectorOperations = { setDisallowedPaths, setFileSelectorCallback, showFileSelector, fetchPageFromPath };
        const favoriteFile = (files: File[]) => updateFiles(favoriteFileFromPage(page, files, Cloud));
        const fileOperations: FileOperation[] = [
            {
                text: "Rename", onClick: (files: File[]) => updateFiles(startRenamingFiles(files, page)),
                disabled: (files: File[]) => false, icon: "edit", color: undefined
            },
            ...AllFileOperations(true, fileSelectorOperations, refetch, this.props.history)
        ];
        return (
            <Grid>
                <Grid.Column computer={13} tablet={16}>
                    <Grid.Row>
                        <Responsive
                            as={ContextButtons}
                            maxWidth={991}
                            createFolder={() => this.newFolder()}
                            currentPath={path}
                            showUploader={this.props.showUploader}
                        />
                        <BreadCrumbs currentPath={path} navigate={(newPath) => navigate(newPath)} homeFolder={Cloud.homeFolder} />
                    </Grid.Row>
                    <Pagination.List
                        loading={loading}
                        errorMessage={error}
                        onErrorDismiss={this.props.dismissError}
                        customEmptyPage={(<Header.Subheader content="No files in current folder" />)}
                        pageRenderer={(page) => (
                            <FilesTable
                                onFavoriteFile={favoriteFile}
                                fileOperations={fileOperations}
                                sortFiles={(sortOrder: SortOrder, sortBy: SortBy) => fetchFiles(path, page.itemsPerPage, page.pageNumber, sortOrder, sortBy)}
                                sortingIcon={(name: SortBy) => UF.getSortingIcon(sortBy, sortOrder, name)}
                                sortOrder={sortOrder}
                                sortingColumns={[leftSortingColumn, rightSortingColumn]}
                                refetchFiles={() => refetch()}
                                onDropdownSelect={(sortOrder: SortOrder, sortBy: SortBy, index: number) => fetchFiles(path, page.itemsPerPage, page.pageNumber, sortOrder, sortBy, index)}
                                masterCheckbox={checkbox}
                                onRenameFile={this.onRenameFile}
                                files={page.items}
                                sortBy={sortBy}
                                onCheckFile={(checked: boolean, file: File) => checkFile(checked, file.path)}
                            />
                        )}
                        onRefresh={refetch}
                        onItemsPerPageChanged={(pageSize) => fetchFiles(path, pageSize, 0, sortOrder, sortBy)}
                        page={page}
                        onPageChanged={(pageNumber: number) => fetchFiles(path, page.itemsPerPage, pageNumber, sortOrder, sortBy)}
                    />
                </Grid.Column>
                <Responsive as={Grid.Column} computer={3} minWidth={992}>
                    <ContextBar
                        showUploader={this.props.showUploader}
                        fileOperations={fileOperations}
                        files={selectedFiles}
                        createFolder={() => this.newFolder()}
                    />
                </Responsive>
                <FileSelectorModal
                    show={this.props.fileSelectorShown}
                    onHide={() => this.props.showFileSelector(false)}
                    path={this.props.fileSelectorPath}
                    fetchFiles={(path, pageNumber, itemsPerPage) => this.props.fetchSelectorFiles(path, pageNumber, itemsPerPage)}
                    loading={this.props.fileSelectorLoading}
                    errorMessage={this.props.fileSelectorError}
                    onErrorDismiss={() => this.props.onFileSelectorErrorDismiss()}
                    onlyAllowFolders
                    canSelectFolders
                    page={this.props.fileSelectorPage}
                    setSelectedFile={this.props.fileSelectorCallback}
                    disallowedPaths={this.props.disallowedPaths}
                />
            </Grid>);
    }
}

export const FilesTable = ({
    files, masterCheckbox, sortingIcon, sortFiles, onRenameFile, onCheckFile, sortingColumns, onDropdownSelect,
    fileOperations, sortOrder, onFavoriteFile, sortBy
}: FilesTableProps) => (
        <Table unstackable basic="very">
            <FilesTableHeader
                onDropdownSelect={onDropdownSelect}
                sortOrder={sortOrder}
                sortingColumns={sortingColumns}
                masterCheckbox={masterCheckbox}
                toSortingIcon={sortingIcon}
                sortFiles={sortFiles}
                sortBy={sortBy}
            />
            <Table.Body>
                {files.map((file: File, i: number) => (
                    <Table.Row className="file-row" key={i}>
                        <FilenameAndIcons
                            file={file}
                            onFavoriteFile={onFavoriteFile}
                            hasCheckbox={masterCheckbox != null}
                            onRenameFile={onRenameFile}
                            onCheckFile={(checked: boolean) => onCheckFile(checked, file)}
                        />
                        <Responsive as={Table.Cell} minWidth={768} content={sortingColumns ? UF.sortingColumnToValue(sortingColumns[0], file) : dateToString(file.modifiedAt)} />
                        <Responsive as={Table.Cell} minWidth={768} content={sortingColumns ? UF.sortingColumnToValue(sortingColumns[1], file) : UF.getOwnerFromAcls(file.acl)} />
                        <Table.Cell>
                            <Dropdown direction="left" icon="ellipsis horizontal">
                                <Dropdown.Menu>
                                    <FileOperations files={[file]} fileOperations={fileOperations} As={Dropdown.Item} />
                                </Dropdown.Menu>
                            </Dropdown>
                        </Table.Cell>
                    </Table.Row>)
                )}
            </Table.Body>
        </Table>
    );

const ResponsiveTableColumn = ({ asDropdown, iconName, onSelect = (sO: SortOrder, sB: SortBy) => null, isSortedBy, currentSelection, sortOrder, minWidth = undefined }: ResponsiveTableColumnProps) => (
    <Responsive minWidth={minWidth} as={Table.HeaderCell}>
        <SortByDropdown isSortedBy={isSortedBy} onSelect={onSelect} asDropdown={asDropdown} currentSelection={currentSelection} sortOrder={sortOrder} />
        <Icon className="float-right" name={iconName} />
    </Responsive>
)

const toSortOrder = (sortBy: SortBy, lastSort: SortBy, sortOrder: SortOrder) =>
    sortBy === lastSort ? (sortOrder === SortOrder.ASCENDING ? SortOrder.DESCENDING : SortOrder.ASCENDING) : SortOrder.ASCENDING;

const FilesTableHeader = ({ toSortingIcon = () => undefined, sortFiles = () => null, sortOrder, masterCheckbox, sortingColumns, onDropdownSelect, sortBy }: FilesTableHeaderProps) => (
    <Table.Header>
        <Table.Row>
            <Table.HeaderCell className="filename-row" onClick={() => sortFiles(toSortOrder(SortBy.PATH, sortBy, sortOrder), SortBy.PATH)}>
                {masterCheckbox}
                Filename
                <Icon className="float-right" name={toSortingIcon(SortBy.PATH)} />
            </Table.HeaderCell>
            {sortingColumns.map((sC, i) => (
                <ResponsiveTableColumn
                    key={i}
                    isSortedBy={sC === sortBy}
                    minWidth={768}
                    onSelect={(sortOrder: SortOrder, sortBy: SortBy) => { if (!!onDropdownSelect) onDropdownSelect(sortOrder, sortBy, i) }}
                    currentSelection={sC}
                    sortOrder={sortOrder}
                    asDropdown={!!onDropdownSelect}
                    iconName={toSortingIcon(sC)}
                />
            ))}
            <Table.HeaderCell />
        </Table.Row>
    </Table.Header>
);

const SortByDropdown = ({ currentSelection, sortOrder, onSelect, asDropdown, isSortedBy }: SortByDropdownProps) => asDropdown ? (
    <Dropdown simple text={UF.prettierString(currentSelection)}>
        <Dropdown.Menu>
            <Dropdown.Item text={UF.prettierString(SortOrder.ASCENDING)} onClick={() => onSelect(SortOrder.ASCENDING, currentSelection)} disabled={sortOrder === SortOrder.ASCENDING && isSortedBy} />
            <Dropdown.Item text={UF.prettierString(SortOrder.DESCENDING)} onClick={() => onSelect(SortOrder.DESCENDING, currentSelection)} disabled={sortOrder === SortOrder.DESCENDING && isSortedBy} />
            <Dropdown.Divider />
            {Object.keys(SortBy).filter(it => it !== currentSelection).map((sortByKey: SortBy, i) => (
                <Dropdown.Item key={i} onClick={() => onSelect(sortOrder, sortByKey)} text={UF.prettierString(sortByKey)} />
            ))}
        </Dropdown.Menu>
    </Dropdown>) : <>{UF.prettierString(currentSelection)}</>;

const ContextBar = ({ files, ...props }: ContextBarProps) => (
    <div>
        <ContextButtons showUploader={props.showUploader} createFolder={props.createFolder} />
        <Divider />
        <FileOptions files={files} {...props} />
    </div>
);

const ContextButtons = ({ createFolder, showUploader }: ContextButtonsProps) => (
    <div>
        <Button color="blue" className="context-button-margin" fluid onClick={() => showUploader()} content="Upload Files" />
        <Button basic className="context-button-margin" fluid onClick={() => createFolder()} content="New folder" />
        <Button as={Link} to={`/filesearch`} basic className="context-button-margin" fluid content="Advanced Search" color="green" />
    </div>
);

const PredicatedCheckbox = ({ predicate, checked, onClick }) =>
    predicate ? (
        <Checkbox
            checked={checked}
            type="checkbox"
            className="hidden-checkbox checkbox-margin"
            onClick={onClick}
        />
    ) : null;

const PredicatedFavorite = ({ predicate, item, onClick }) =>
    predicate ? (
        <Icon
            color="blue"
            name={item.favorited ? "star" : "star outline"}
            className={`${item.favorited ? "" : "file-data"} favorite-padding`}
            onClick={onClick}
        />
    ) : null;

const GroupIcon = ({ isProject }: { isProject: boolean }) => isProject ? (<Icon className="group-icon-padding" name="users" />) : null;

const FileLink = ({ file, children }) => {
    if (isDirectory(file)) {
        return (<Link to={`/files/${file.path}`}>{children}</Link>);
    } else if (previewSupportedExtension(file.path)) {
        return (<Link to={`/filepreview/${file.path}`}>{children}</Link>);
    } else {
        return (<>{children}</>);
    }
}

function FilenameAndIcons({ file, size = "big", onRenameFile = () => null, onCheckFile = () => null, hasCheckbox = false, onFavoriteFile = () => null }: FilenameAndIconsProps) {
    const fileName = getFilenameFromPath(file.path);
    const checkbox = <PredicatedCheckbox predicate={hasCheckbox} checked={file.isChecked} onClick={(_, { checked }) => onCheckFile(checked)} />
    const icon = (
        <FileIcon
            color={isDirectory(file) ? "blue" : "grey"}
            name={UF.iconFromFilePath(file.path, file.fileType, Cloud.homeFolder)}
            size={size} link={file.link}
        />
    );
    const nameLink = <FileLink file={file}>{icon}{fileName}</FileLink>;
    return file.beingRenamed ?
        <Table.Cell className="table-cell-padding-left">
            <Input
                defaultValue={fileName}
                onKeyDown={(e) => { if (!!onRenameFile) onRenameFile(e.keyCode, file, e.target.value) }}
                autoFocus
                transparent
                fluid
            >
                {checkbox}
                {icon}
                <input />
                <Button content="Cancel" size="small" color="red" basic onClick={() => onRenameFile(KeyCode.ESC, file, "")} />
            </Input>

        </Table.Cell> :
        <Table.Cell className="table-cell-padding-left">
            {checkbox}
            {nameLink}
            <GroupIcon isProject={isProject(file)} />
            <PredicatedFavorite predicate={!!onFavoriteFile && !file.path.startsWith(`${Cloud.homeFolder}Favorites`)} item={file} onClick={() => onFavoriteFile([file])} />
        </Table.Cell>
};

const FileOptions = ({ files, fileOperations }: FileOptionsProps) => files.length ? (
    <div>
        <Header as="h3">{toFileText(files)}</Header>
        <FileOperations files={files} fileOperations={fileOperations} As={Button} fluid basic />
    </div>
) : null;

export const FileOperations = ({ files, fileOperations, As, fluid, basic }) => files.length && fileOperations.length ?
    fileOperations.map((fileOp, i) => {
        let operation = fileOp;
        if ((fileOp as PredicatedOperation).predicate) {
            operation = fileOp as PredicatedOperation
            operation = operation.predicate(files, Cloud) ? operation.onTrue : operation.onFalse;
        }
        operation = operation as Operation;
        return !operation.disabled(files, Cloud) ? (
            <As
                key={i}
                content={operation.text}
                icon={operation.icon}
                color={operation.color}
                className="context-button-margin"
                onClick={() => (operation as Operation).onClick(files, Cloud)}
                fluid={fluid}
                basic={basic}
            />
        ) : null;
    }) : null;

const mapStateToProps = ({ files }: ReduxObject): FilesStateProps => {
    const { page, loading, path, fileSelectorPage, fileSelectorPath, sortBy, sortOrder, fileSelectorShown,
        fileSelectorCallback, disallowedPaths, fileSelectorLoading, error, fileSelectorError, sortingColumns } = files;
    const favFilesCount = page.items.filter(file => file.favorited).length; // HACK to ensure changes to favorites are rendered.
    const checkedFilesCount = page.items.filter(file => file.isChecked).length; // HACK to ensure changes to file checkings are rendered.
    const renamingCount = page.items.filter(file => file.beingRenamed).length;
    const fileCount = page.items.length;
    return {
        error, fileSelectorError, page, loading, path, checkedFilesCount, favFilesCount, fileSelectorPage, fileSelectorPath,
        fileSelectorShown, fileSelectorCallback, disallowedPaths, sortOrder, sortBy, fileCount, fileSelectorLoading,
        leftSortingColumn: sortingColumns[0], rightSortingColumn: sortingColumns[1], renamingCount
    }
};

const mapDispatchToProps = (dispatch: Dispatch): FilesOperations => ({
    prioritizeFileSearch: () => dispatch(setPrioritizedSearch("files")),
    onFileSelectorErrorDismiss: () => dispatch(Actions.setFileSelectorError(undefined)),
    dismissError: () => dispatch(Actions.setErrorMessage()),
    fetchFiles: async (path: string, itemsPerPage: number, pageNumber: number, sortOrder: SortOrder, sortBy: SortBy, index?: number) => {
        dispatch(Actions.updatePath(path));
        dispatch(Actions.setLoading(true));
        if (index != null) dispatch(Actions.setSortingColumn(sortBy, index));
        dispatch(await Actions.fetchFiles(path, itemsPerPage, pageNumber, sortOrder, sortBy));
    },
    fetchPageFromPath: async (path: string, itemsPerPage: number, sortOrder: SortOrder, sortBy: SortBy) => {
        dispatch(Actions.setLoading(true));
        dispatch(await Actions.fetchPageFromPath(path, itemsPerPage, sortOrder, sortBy));
    },
    updatePath: (path: string) => dispatch(Actions.updatePath(path)),
    fetchSelectorFiles: async (path: string, pageNumber: number, itemsPerPage: number) => dispatch(await Actions.fetchFileselectorFiles(path, pageNumber, itemsPerPage)),
    showFileSelector: (open: boolean) => dispatch(Actions.fileSelectorShown(open)),
    setFileSelectorCallback: (callback) => dispatch(Actions.setFileSelectorCallback(callback)),
    checkFile: (checked: boolean, path: string) => dispatch(Actions.checkFile(checked, path)),
    setPageTitle: () => dispatch(updatePageTitle("Files")),
    updateFiles: (page: Page<File>) => dispatch(Actions.updateFiles(page)),
    checkAllFiles: (checked: boolean) => dispatch(Actions.checkAllFiles(checked)),
    setDisallowedPaths: (disallowedPaths: string[]) => dispatch(Actions.setDisallowedPaths(disallowedPaths)),
    showUploader: () => dispatch(setUploaderVisible(true)),
    setUploaderCallback: (callback) => dispatch(setUploaderCallback(callback))
});

export default connect(mapStateToProps, mapDispatchToProps)(Files);