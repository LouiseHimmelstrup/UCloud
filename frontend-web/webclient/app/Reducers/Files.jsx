export const RECEIVE_FILES = "RECEIVE_FILES";
export const SET_FAVORITE = "SET_FAVORITE";
export const UPDATE_FILES_PER_PAGE = "UPDATE_FILES_PER_PAGE";
export const UPDATE_FILES = "UPDATE_FILES";
export const SET_LOADING = "SET_LOADING";
export const UPDATE_PATH = "UPDATE_PATH";
export const TO_PAGE = "TO_PAGE";
export const UPDATE_FILES_INFO_PATH = "UPDATE_FILES_INFO_PATH";

const files = (state = [], action) => {
    switch (action.type) {
        case RECEIVE_FILES: {
            return { ...state, files: action.files, loading: false };
        }
        case UPDATE_FILES_PER_PAGE: {
            return { ...state, files: action.files, filesPerPage: action.filesPerPage, currentFilesPage: 0 };
        }
        case UPDATE_FILES: {
            return { ...state, files: action.files };
        }
        case SET_LOADING: {
            return { ...state, loading: action.loading };
        }
        case UPDATE_PATH: {
            return { ...state, path: action.path };
        }
        case TO_PAGE: {
            return { ...state, currentFilesPage: action.pageNumber };
        }
        default: {
            return state;
        }
    }
}

export default files;