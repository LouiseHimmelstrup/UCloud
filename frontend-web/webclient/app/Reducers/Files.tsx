export const RECEIVE_FILES = "RECEIVE_FILES";
export const UPDATE_FILES = "UPDATE_FILES";
export const SET_FILES_LOADING = "SET_FILES_LOADING";
export const UPDATE_PATH = "UPDATE_PATH";
export const UPDATE_FILES_INFO_PATH = "UPDATE_FILES_INFO_PATH";
export const SET_FILES_SORTING_COLUMN = "SET_FILES_SORTING_COLUMN";
export const FILE_SELECTOR_SHOWN = "FILE_SELECTOR_SHOWN";
export const RECEIVE_FILE_SELECTOR_FILES = "RECEIVE_FILE_SELECTOR_FILES";
export const SET_FILE_SELECTOR_LOADING = "SET_FILE_SELECTOR_LOADING";
export const SET_FILE_SELECTOR_CALLBACK = "SET_FILE_SELECTOR_CALLBACK";
export const SET_DISALLOWED_PATHS = "SET_DISALLOWED_PATHS";

const files = (state: any = {}, action) => {
    switch (action.type) {
        case RECEIVE_FILES: {
            return { ...state, page: action.page, loading: false, fileSelectorFiles: action.page.items, fileSelectorPath: action.path, sortOrder: action.sortOrder, sortBy: action.sortBy };
        }
        case UPDATE_FILES: {
            return { ...state, page: action.page };
        }
        case SET_FILES_LOADING: {
            return { ...state, loading: action.loading };
        }
        case UPDATE_PATH: {
            return { ...state, path: action.path };
        }
        case FILE_SELECTOR_SHOWN: {
            return { ...state, fileSelectorShown: action.state };
        }
        case RECEIVE_FILE_SELECTOR_FILES: {
            return { ...state, fileSelectorFiles: action.files, fileSelectorPath: action.path, fileSelectorLoading: false };
        }
        case SET_FILE_SELECTOR_LOADING: {
            return { ...state, fileSelectorLoading: true };
        }
        case SET_FILE_SELECTOR_CALLBACK: {
            return { ...state, fileSelectorCallback: action.callback };
        }
        case SET_DISALLOWED_PATHS: {
            return { ...state, disallowedPaths: action.paths }
        }
        default: {
            return state;
        }
    }
}

export default files;