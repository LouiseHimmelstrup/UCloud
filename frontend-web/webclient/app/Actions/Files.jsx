import {Cloud} from "../../authentication/SDUCloudObject"
import { RECEIVE_FILES, UPDATE_FILES_PER_PAGE, UPDATE_FILES, SET_LOADING, UPDATE_PATH, TO_PAGE } from "../Reducers/Files"

/*
* Creates a promise to fetch files. Sorts the files based on sorting function passed,
* and implicitely sets @filesLoading to false in the reducer when the files are fetched.
*/
export const fetchFiles = (path, sorting, sortAscending) => 
  Cloud.get(`files?path=${path}`).then((res) => {
      let files = res.response;
      files.forEach(file => file.isChecked = false);
      if (sorting) {
        files = sorting(files, sortAscending);
      }
      return receiveFiles(files);
  });

export const toPage = (pageNumber) => ({
  type: TO_PAGE,
  pageNumber: pageNumber
});

/*
* Updates the files stored. 
* Intended for use when sorting the files, checking or favoriting, for instance.
*/
export const updateFiles = (files) => {
  return {
    type: UPDATE_FILES,
    files
  }
};

export const setLoading = (loading) => {
  return {
    type: SET_LOADING,
    loading
  }
};

export const updatePath = (newPath) => ({
    type: UPDATE_PATH,
    path: newPath
});

export const updateFilesPerPage = (filesPerPage, files) => {
  files.forEach(file => file.isChecked = false);
  return {
    type: UPDATE_FILES_PER_PAGE,
    filesPerPage,
    files
  }
};

const receiveFiles = (files, path) => ({
  type: RECEIVE_FILES,
  files,
  path
});