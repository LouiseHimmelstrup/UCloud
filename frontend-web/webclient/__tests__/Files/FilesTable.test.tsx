import * as React from "react";
import FilesTable from "../../app/Files/FilesTable";
import { SortOrder, SortBy } from "../../app/Files";

describe("FilesTable", () => {
    test("Mount", () => {
        const filesTable = <FilesTable 
            sortOrder={SortOrder.ASCENDING}
            sortBy={SortBy.PATH}
            sortingColumns={[SortBy.ACL, SortBy.CREATED_AT]}
            files={[]}
            refetchFiles={() => undefined}
            sortFiles={() => undefined}
            onCheckFile={() => undefined}
            fileOperations={[]}
        />;
        expect(filesTable).toMatchSnapshot();
    })
});