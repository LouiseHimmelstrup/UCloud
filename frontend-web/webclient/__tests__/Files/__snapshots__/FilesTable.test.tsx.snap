// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilesTable Mount 1`] = `
<Memo(ConnectFunction)
  fileOperations={Array []}
  files={Array []}
  onCheckFile={[Function]}
  refetchFiles={[Function]}
  sortBy="path"
  sortFiles={[Function]}
  sortOrder="ASCENDING"
  sortingColumns={
    Array [
      "acl",
      "createdAt",
    ]
  }
/>
`;
